osu.Game/Rulesets/Objects/Pooling/HitObjectEntryManager.cs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+
+namespace osu.Game.Rulesets.Objects.Pooling
+{
+    /// <summary>
+    /// Manages a mapping between <see cref="HitObject"/> and <see cref="HitObjectLifetimeEntry"/>
+    /// </summary>
+    internal class HitObjectEntryManager
+    {
+        /// <summary>
+        /// All entries, including entries of the nested hit objects.
+        /// </summary>
+        public IEnumerable<HitObjectLifetimeEntry> AllEntries => entryMap.Values;
+
+        /// <summary>
+        /// Invoked when a new <see cref="HitObjectLifetimeEntry"/> is added to this <see cref="HitObjectEntryManager"/>..
+        /// The second parameter of the event is the parent hit object.
+        /// </summary>
+        public event Action<HitObjectLifetimeEntry, HitObject?>? OnEntryAdded;
+
+        /// <summary>
+        /// Invoked when a <see cref="HitObjectLifetimeEntry"/> is removed from this <see cref="HitObjectEntryManager"/>.
+        /// The second parameter of the event is the parent hit object.
+        /// </summary>
+        public event Action<HitObjectLifetimeEntry, HitObject?>? OnEntryRemoved;
+
+        /// <summary>
+        /// Provides the reverse mapping of <see cref="HitObjectLifetimeEntry.HitObject"/> for each entry.
+        /// </summary>
+        private readonly Dictionary<HitObject, HitObjectLifetimeEntry> entryMap = new Dictionary<HitObject, HitObjectLifetimeEntry>();
+
+        /// <summary>
+        /// Stores the parent hit object for entries of the nested hit objects.
+        /// A <c>null</c> is stored for entries of the top-level hit objects.
+        /// </summary>
+        /// <remarks>
+        /// The parent hit object of a pooled hit object may be non-pooled.
+        /// In that case, no corresponding <see cref="HitObjectLifetimeEntry"/> is stored in this <see cref="HitObjectEntryManager"/>.
+        /// </remarks>
+        private readonly Dictionary<HitObjectLifetimeEntry, HitObject?> parentMap = new Dictionary<HitObjectLifetimeEntry, HitObject?>();
+
+        /// <summary>
+        /// Stores the list of entries managed by this <see cref="HitObjectEntryManager"/> for each hit object managed by this <see cref="HitObjectEntryManager"/>.
+        /// </summary>
+        private readonly Dictionary<HitObject, List<HitObjectLifetimeEntry>> childrenMap = new Dictionary<HitObject, List<HitObjectLifetimeEntry>>();
+
+        public void Add(HitObjectLifetimeEntry entry, HitObject? parent)
+        {
+            if (parentMap.ContainsKey(entry))
+                throw new InvalidOperationException($@"The {nameof(HitObjectLifetimeEntry)} is already added to this {nameof(HitObjectEntryManager)}.");
```

Comment:
Any reason for checking `parentMap` here (and in `Remove()`), instead of `entryMap`? Perhaps you could assert that `parentMap` doesn't contain it, but I feel like `entryMap` explicitly saying "each entry" (understanding to mean all entries) makes it more explicit as to what's going on.